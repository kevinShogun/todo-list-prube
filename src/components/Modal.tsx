
interface ModalProps {
    children: React.ReactNode;
    id?: string;
}

export const Modal = (
    { children, id = 'my_modal_2' }: ModalProps,
) => {

    return (
        <dialog id={id} className="modal">
            <div className="modal-box">
                {children}
            </div>
            <form method="dialog" className="modal-backdrop">
                <button>close</button>
            </form>
        </dialog>
    )
}
