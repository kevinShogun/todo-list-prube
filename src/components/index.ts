export { TodoScreen } from "./TodoScreen";
export { SideBar } from "./SideBar";

