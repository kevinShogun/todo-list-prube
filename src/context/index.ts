export { TodoProvider } from "./TodoProvider";
export { TodoReducer } from "./TodoReducer";
export { TodoContext } from "./TodoContext";


