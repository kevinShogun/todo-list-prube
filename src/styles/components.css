.button-text{
    @apply font-semibold text-lg md:text-xl lg:text-2xl;
}

.color-circle{
    @apply w-5 h-5 shadow-md rounded-full cursor-pointer mx-1 md:w-7 md:h-7 tooltip;
}

.input-text{
    @apply input input-bordered w-full mt-2;
}

form{
    @apply flex flex-col gap-3 select-none;
}
.textarea-form{
    @apply textarea textarea-bordered w-full  mt-2 resize-none min-h-32;    
}
