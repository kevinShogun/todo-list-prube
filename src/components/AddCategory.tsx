import { useState, useContext, useEffect } from 'react'
import { v4 as uuid } from 'uuid';
import { TodoContext } from '../context';
import { toggleModal } from '../helper/modalAction';
import { Category } from '../interfaces'
import { COLORS } from '../constants/colors';
import { Modal } from './Modal'
import '../styles/components.css'


/**
 * @description This code defines a React functional component called AddCategory
 * that is responsible for adding and updating categories. 
 * It uses the useState and useContext hooks from React to manage the state and access the context. 
 * The component renders a button that opens a modal dialog when clicked. 
 * Inside the modal, the user can enter a category name and select a color from a predefined list. 
 * The component also listens for changes in the currentCategory context variable and 
 * updates the input fields accordingly.
 * 
 * @returns  {JSX.Element} The rendered todo item.
 */
export const AddCategory = () => {

    const [categoryName, setCategoryName] = useState('');
    const [color, setColor] = useState('');

    const { addCategory, currentCategory, replaceCategory } = useContext(TodoContext);

    useEffect(() => {
        if (currentCategory) {
            setCategoryName(currentCategory.name);
            setColor(currentCategory.color);
            toggleModal({ value: true });
        }
    }, [currentCategory])


    const handleCategory = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault()

        if (categoryName.trim() === '' || color.trim() === '') {
            return
        }

        const newCategory: Category = {
            id: uuid(),
            name: categoryName,
            color
        }

        if (currentCategory) {
            const updatedCategory = { ...currentCategory, name: categoryName, color }
            replaceCategory(updatedCategory)
        }else{
            addCategory(newCategory);
        }
        
        setCategoryName('');
        setColor('');

        toggleModal({ value: false });
    }

    return (
        <div>
            <button className="btn btn-ghost"
                onClick={() => {
                    toggleModal({ value: true });
                }}
            >
                Agregar una categoria
                <i className='bx bx-plus'></i>
            </button>
            <Modal>
                <h2 className="button-text">
                    Agregar una categoria
                </h2>
                <div className="divider" />
                <form method="dialog" className="modal-form" onSubmit={handleCategory}>
                    <div className="form-control">
                        <label htmlFor="category">Categoria: </label>
                        <input type="text" id="category" className='input-text'
                            value={categoryName}
                            onChange={(e) => setCategoryName(e.target.value)}
                        />
                    </div>

                    <div className="form-control">
                        <label htmlFor="color">Color: </label>
                        <div className='flex gap-3'>
                            {COLORS.map((c, index) => (
                                <div
                                    key={index}
                                    className={`color-circle ${color === c && 'border-2 border-black'}`}
                                    style={{ backgroundColor: c }}
                                    onClick={() => setColor(c)}
                                />

                            ))}

                        </div>

                    </div>
                    <button type="submit" className="btn btn-primary">
                        {currentCategory ? 'Actualizar' : 'Agregar'}
                    </button>
                </form>
            </Modal>
        </div>
    )
}
