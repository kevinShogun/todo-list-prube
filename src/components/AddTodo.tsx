import { useState, useContext, useEffect } from 'react';
import { v4 as uuid } from 'uuid';
import { TodoContext } from '../context';
import { Category, Todo } from '../interfaces';
import { toggleModal } from "../helper/modalAction"
import { Modal } from "./Modal"
import '../styles/components.css'

type NewTodo = {
    title: string;
    body: string;
    categories: Category[];
}

/**
 * @description This code defines a React component called AddTodo that is responsible for adding a new todo item. 
 * It uses the useState and useContext hooks from React to manage the state and access the context. 
 * The component renders a button that opens a modal dialog when clicked. 
 * Inside the modal, the user can enter a title, description, and select categories for the new todo item. 
 * The component also handles updating an existing todo item if one is provided through the context.
 * @returns {JSX.Element}
 */

export const AddTodo = () => {

    const [todo, settodo] = useState<NewTodo>({
        title: '',
        body: '',
        categories: []
    });
    const { categories, addTodo, replaceTodo, currentTodo} = useContext(TodoContext);


    useEffect(() => {
        if (currentTodo) {
            settodo({
                title: currentTodo.title,
                body: currentTodo.body,
                categories: currentTodo.categories
            });
            toggleModal({ value: true, id: 'add_todo'});
        }
    }, [currentTodo])


    const setCategories = (c: Category) => {

        if(todo.categories.includes(c)){
            const newCategories = todo.categories.filter((cat) => cat.id !== c.id);
            return settodo({ ...todo, categories: newCategories })
        }else{
            return settodo({ 
                ...todo, 
                categories: [...todo.categories, c]
            })
        }
    }

    const handleAddTodo = () => {
        const newTodo:Todo = {
            ...todo,
            id: uuid(),
        }

        if(currentTodo){
            const updatedTodo = {
                ...currentTodo,
                title: todo.title,
                body: todo.body,
                categories: todo.categories
            }
            replaceTodo(updatedTodo);
        }else{
            addTodo(newTodo);
        }

        settodo({
            title: '',
            body: '',
            categories: []
        });
        toggleModal({ value: false, id: 'add_todo'});

    }

    return (
        <div>
            <button className="btn btn-ghost"
                onClick={() => {
                    toggleModal({ value: true, id: 'add_todo' });
                }}
            >
                Agregar una tarea
                <i className='bx bx-plus'></i>
            </button>
            <br /> <br />
            <Modal
                id="add_todo"
            >
                <h2 className="button-text">
                    Nueva tarea
                </h2>
                <div className="divider" />
                <form method="dialog" className="modal-form" onSubmit={handleAddTodo}>
                    <div className="form-control">
                        <label htmlFor="title">Título: </label>
                        <input type="text" id="title" className='input-text'
                            value={todo.title}
                            onChange={(e) => settodo({ ...todo, title: e.target.value })}
                        />
                    </div>

                    <div className="form-control">
                        <label htmlFor="description">Descripción: </label>
                        <textarea
                            id="description"
                            value={todo.body}
                            className='textarea-form'
                            onChange={(e) => settodo({ ...todo, body: e.target.value })}
                        ></textarea>
                    </div>
                    <div className="form-control">
                        <label htmlFor="color">Categorías: </label>
                        <div className='flex gap-3 mt-2'>
                            {categories?.map((c, index) => (
                                <div
                                    key={index}
                                    className={`color-circle ${todo.categories.includes(c) ? 'border-2 border-black' : ''}`}
                                    data-tip={c.name}
                                    style={{ backgroundColor: c.color }}
                                    onClick={() =>  setCategories(c) }
                                />

                            ))}

                        </div>

                    </div>

                    <button type="submit" className="btn btn-primary">
                        {currentTodo ? 'Actualizar' : 'Agregar'}
                    </button>
                </form>
            </Modal>
        </div>
    )
}
