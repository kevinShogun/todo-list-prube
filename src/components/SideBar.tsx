import { useContext } from 'react'
import { MagicMotion } from "react-magic-motion";
import { AddCategory } from "./AddCategory"
import { TodoContext } from "../context";

export const SideBar = () => {

  const { categories, setCurrentCategory, removeCategory } = useContext(TodoContext);

  return (
    <aside className="sidebar">
      <h2 className="font-semibold text-2xl md:text-3xl lg:text-5xl">
        todo
      </h2>

      <MagicMotion>
        <div>
          <ul className="mt-5">
            {categories.map((category) => (
              <li key={category.id} className="list-item">
                <div className='flex w-3/4 items-center gap-2'>
                  <span className="category-color" style={{ backgroundColor: category.color }} />
                  <span className='list-item'>{category.name}</span>
                </div>
                <div className="dropdown dropdown-end">
                  <div tabIndex={0} role="button">
                    <i className='bx bx-dots-horizontal-rounded'></i>
                  </div>
                  <ul tabIndex={0} className="dropdown-item">
                    <li>
                      <a onClick={() => setCurrentCategory(category)}> Editar </a>
                    </li>
                    <li>
                      <a onClick={() => removeCategory(category.id)}> Eliminar </a>
                    </li>
                  </ul>
                </div>
              </li>
            ))}
          </ul>
        </div>
      </MagicMotion>

      <div className="divider" />
      <AddCategory />
      <img
        src='./undraw_to_do_re_jaef.svg'
        alt="note list"
        className="w-3/4 mx-auto mt-5 hidden lg:block"
      />
    </aside>
  )
}
/***
 * Edad de 25 a 35 a√±os
talentohumano.ni@grupolgb.com
whatsapp 76221311
 */
