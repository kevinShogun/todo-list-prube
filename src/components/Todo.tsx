import { FC, useContext } from 'react';
import { Todo as ITodo } from "../interfaces"
import { TodoContext } from '../context';
import '../styles/todoStyle.css';
import '../styles/components.css';


interface TodoProps {
    todo: ITodo;
}
/**
 * Renders a single todo item.
 * @param {TodoProps} props - The props object containing the todo item.
 * @param {ITodo} props.todo - The todo item to be rendered.
 * @returns {JSX.Element} The rendered todo item.
 */
export const Todo: FC<TodoProps> = ({ todo }) => {
    
    const { removeTodo, setCurrentTodo, toggleTodo} = useContext(TodoContext);

    return (
        <div className="todo-container">
            <div className="todo-dropdown">
                <div tabIndex={0} role="button">
                    <i className='bx bx-dots-horizontal-rounded'></i>
                </div>
                <ul tabIndex={0} className="dropdown-content z-[1] menu p-2 shadow bg-base-100 rounded-box w-52">
                    <li><a
                        className={`${todo.done && 'hidden'}`}
                        onClick={() => setCurrentTodo(todo)}
                    >Editar </a></li>
                    <li><a
                        onClick={() => removeTodo(todo.id)}
                    >Eliminar</a></li>
                </ul>
            </div>
            <h3 className={`todo-title ${todo.done && 'line-through text-[#6d6a5f]'}` }
            > { todo.title } </h3>
            <p className="todo-text">{todo.body}</p>
            <div className="container-flex">
                <div className='flex gap-3 mt-2'>
                    {todo.categories.map((c, index) => (
                        <div
                            key={index}
                            className={`color-circle ${todo.categories.includes(c) && 'border-2 border-black'}`}
                            data-tip={c.name}
                            style={{ backgroundColor: c.color }}
                        />
                    ))}
                </div>

                <label className="flex items-center space-x-2 mt-3">
                    <input type="checkbox" className="checkbox rounded-md text-primary-500"
                        checked={todo.done}
                        onChange={() => { toggleTodo(todo.id)}}
                    />
                    <span className="todo-bottom-text">Completado</span>
                </label>
            </div>

        </div>
    )
}
