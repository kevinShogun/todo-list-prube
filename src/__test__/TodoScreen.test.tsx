import { fireEvent, render, screen } from "@testing-library/react";
import { TodoScreen } from "../components";
import { TodoContext } from "../context";


const todos = [
    { id: "1", title: "Task 1", body: "This is task 1", done: true, categories: [] },
    { id: "2", title: "Task 2", body: "This is task 2", done: false, categories: [] },
    { id: "3", title: "Task 3", body: "This is task 3", done: false, categories: [] },
];

describe("TodoScreen Tests Suites", () => {


    test("should match snapshot", () => {
        const component = render(<TodoScreen />);
        expect(component).toMatchSnapshot();
    });

    it('should render the component without crashing', () => {
        render(
            <TodoContext.Provider value={{
                todos: todos,
                categories: [],
                currentTodo: null,
                setCurrentTodo: () => { }, // Add the missing properties here
                currentCategory: { id: '1', name: 'Work', color: 'red' },
                addCategory: () => { },
                replaceCategory: () => { },
                addTodo: () => { },
                replaceTodo: () => { },
                removeCategory(id) {
                    this.categories = this.categories.filter((c) => c.id !== id);
                },
                removeTodo: () => { },
                setCurrentCategory(category) {
                    this.currentCategory = category;
                },
                toggleTodo: () => { },
            }}>
                <TodoScreen />
            </TodoContext.Provider>
        );

        // console.log(screen.debug())
        const colorCircles = todos.map((t) => screen.getByTestId(t.title));
        expect(colorCircles.length).toBe(todos.length);

    });

    test('should toggle the done property of a todo when the checkbox is clicked', () => {
        // Arrange
        const toggleTodoMock = vi.fn();

        const TodoContextValue = {
            todos: todos,
            toggleTodo: toggleTodoMock
        };

        render(
            <TodoContext.Provider value={
                {
                    ...TodoContextValue,
                    categories: [],
                    currentTodo: null,
                    setCurrentTodo: () => { }, // Add the missing properties here
                    currentCategory: { id: '1', name: 'Work', color: 'red' },
                    addCategory: () => { },
                    replaceCategory: () => { },
                    addTodo: () => { },
                    replaceTodo: () => { },
                    removeCategory(id) {
                        this.categories = this.categories.filter((c) => c.id !== id);
                    },
                    removeTodo: () => { },
                    setCurrentCategory(category) {
                        this.currentCategory = category;
                    },
                }
            }>
                <TodoScreen />
            </TodoContext.Provider>
        );

        // Act
        const checkbox = screen.getAllByRole('checkbox');
        fireEvent.click(checkbox[0]);
        
        // Assert not called
        expect(toggleTodoMock).not.toHaveBeenCalledTimes(1);
    });
});