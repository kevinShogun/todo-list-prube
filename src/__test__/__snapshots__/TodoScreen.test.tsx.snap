// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoScreen Tests Suites > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full p-3 md:p-5 h-full overflow-y-auto"
      >
        <div>
          <button
            class="btn btn-ghost"
          >
            Agregar una tarea
            <i
              class="bx bx-plus"
            />
          </button>
          <br />
           
          <br />
          <dialog
            class="modal"
            id="add_todo"
          >
            <div
              class="modal-box"
            >
              <h2
                class="button-text"
              >
                Nueva tarea
              </h2>
              <div
                class="divider"
              />
              <form
                class="modal-form"
                method="dialog"
              >
                <div
                  class="form-control"
                >
                  <label
                    for="title"
                  >
                    Título: 
                  </label>
                  <input
                    class="input-text"
                    id="title"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="form-control"
                >
                  <label
                    for="description"
                  >
                    Descripción: 
                  </label>
                  <textarea
                    class="textarea-form"
                    id="description"
                  />
                </div>
                <div
                  class="form-control"
                >
                  <label
                    for="color"
                  >
                    Categorías: 
                  </label>
                  <div
                    class="flex gap-3 mt-2"
                  />
                </div>
                <button
                  class="btn btn-primary"
                  type="submit"
                >
                  Agregar
                </button>
              </form>
            </div>
            <form
              class="modal-backdrop"
              method="dialog"
            >
              <button>
                close
              </button>
            </form>
          </dialog>
        </div>
        <div
          class="grid grid-cols-1 md:grid-cols-2 gap-3 md:gap-5 w-full"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full p-3 md:p-5 h-full overflow-y-auto"
    >
      <div>
        <button
          class="btn btn-ghost"
        >
          Agregar una tarea
          <i
            class="bx bx-plus"
          />
        </button>
        <br />
         
        <br />
        <dialog
          class="modal"
          id="add_todo"
        >
          <div
            class="modal-box"
          >
            <h2
              class="button-text"
            >
              Nueva tarea
            </h2>
            <div
              class="divider"
            />
            <form
              class="modal-form"
              method="dialog"
            >
              <div
                class="form-control"
              >
                <label
                  for="title"
                >
                  Título: 
                </label>
                <input
                  class="input-text"
                  id="title"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="form-control"
              >
                <label
                  for="description"
                >
                  Descripción: 
                </label>
                <textarea
                  class="textarea-form"
                  id="description"
                />
              </div>
              <div
                class="form-control"
              >
                <label
                  for="color"
                >
                  Categorías: 
                </label>
                <div
                  class="flex gap-3 mt-2"
                />
              </div>
              <button
                class="btn btn-primary"
                type="submit"
              >
                Agregar
              </button>
            </form>
          </div>
          <form
            class="modal-backdrop"
            method="dialog"
          >
            <button>
              close
            </button>
          </form>
        </dialog>
      </div>
      <div
        class="grid grid-cols-1 md:grid-cols-2 gap-3 md:gap-5 w-full"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
